botocore/waiter.py

```diff
@@ -272,8 +272,12 @@ def wait(self, **kwargs):
                 if 'Error' in response:
                     # Transition to the failure state, which we can
                     # just handle here by raising an exception.
+                    reason = 'Unexpected error encountered'
+                    if 'Code' in response['Error']:
+                        reason += ': ' + response['Error']['Code']
+
                     raise WaiterError(name=self.name,
-                                      reason='Unexpected error encountered.')
+                                      reason=reason)
```

Comment:
We should probably include the error message.  As your code suggests, there may not always be a code, but there's always a message.  Maybe have it be `"An error {code} occurred: {message}"` where code can get `response['Error'].get('Code', 'Unknown')`